### ISSUE TITLE
SAM doesn't support Fn::If On Policy Templates

### ISSUE THREAD

**Issue by mlaff** [awslabs/serverless-application-model#889]
**Description:**

SAM doesn't support `Fn::If` in `Policies` section of `AWS::Serverless::Function` when managed policies are listed.

Using template: 

```yaml
Conditions:
  StreamNameProvided:
    Fn::Not:
    - Fn::Equals:
      - Ref: StreamName
      - ''

Resources:
  ProcessLogs:
    Properties:
      InlineCode: print('hi')
      Description: Do stuff
      Handler: index.handler
      MemorySize: 128
      Policies:
        - Fn::If:
          - StreamNameProvided
          - KinesisStreamReadPolicy
            StreamName:
              Ref: StreamName
          - AWS::NoValue
        - Statement:
          Action: cloudwatch:PutMetricData
          Effect: Allow
          Resource: '*'
      Runtime: nodejs8.10
      Timeout: 10
    Type: AWS::Serverless::Function
```
we fail with an unhandled exception. But as seen in [874](https://github.com/awslabs/serverless-application-model/issues/874), this (sort of) works with explicit policy statements. `Fn::If` should work with managed policies as well as explicit ones.

**Comment by jiabiao** [comment]
SAM also does not support `Fn::If` in `Events`section of AWS::Serverless::Function.

**Comment by jlhood** [comment]
@jiabiao Thanks for your comment. It's true SAM's support for CloudFormation intrinsic functions is limited. SAM does support intrinsic functions in the `Events` section of `AWS::Serverless::Function`. However, the level of support depends on which event type is being used. [This page](https://github.com/awslabs/serverless-application-model/blob/develop/docs/cloudformation_compatibility.rst#events-properties) specifies intrinsic function support per event type. If there's a specific case within the `Events` section where SAM does not currently support `Fn::If`, but you need it to, you should open a new GitHub issue with a sample template that demonstrates the specific case that you need supported. Thanks!

**Comment by keetonian** [comment]
This was merged with https://github.com/awslabs/serverless-application-model/pull/988 and shipped in v1.14.0.
